src/dstack/_internal/server/background/tasks/process_instances.py

```diff
@@ -88,6 +117,136 @@ async def process_instances() -> None:
+        ).one()
+
+        retry_duration_deadline = instance.created_at.replace(
+            tzinfo=datetime.timezone.utc
+        ) + timedelta(seconds=PROVISIONING_TIMEOUT_SECONDS)
+        if retry_duration_deadline < get_current_datetime():
+            instance.status = InstanceStatus.TERMINATED
+            instance.deleted = True
+            instance.deleted_at = get_current_datetime()
+            instance.termination_reason = "The proivisioning timeout expired"
+            await session.commit()
+            logger.warning(
+                "Failed to start the instance in %s seconds. Terminate instance %s",
+                PROVISIONING_TIMEOUT_SECONDS,
+                instance.name,
+                extra={
+                    "instance_name": instance.name,
+                    "instance_status": InstanceStatus.TERMINATED.value,
+                },
+            )
+            return
+
+        try:
+            remote_details = RemoteConnectionInfo.parse_raw(
+                cast(str, instance.remote_connection_info)
+            )
+
+            # Prepare connection key
+            try:
+                private_string = [
+                    sk.private for sk in remote_details.ssh_keys if sk.private is not None
+                ][0]
+            except IndexError:
+                logger.error("There are no ssh private key")
+                raise BackendError()
+            pkey = rsa_pkey_from_str(convert_pkcs8_to_pem(private_string))
+
+            with get_paramiko_connection(
+                remote_details.ssh_user, remote_details.host, remote_details.port, pkey
+            ) as client:
+                logger.info(f"connected to {remote_details.ssh_user} {remote_details.host}")
+
+                runner_build = get_dstack_runner_version()
+
+                # Upload envs
+                shim_envs = get_shim_env(
+                    runner_build, authorized_keys=[sk.public for sk in remote_details.ssh_keys]
+                )
+                upload_envs(client, DSTACK_WORKING_DIR, shim_envs)
+                logger.debug("The dstack-shim environemnt variables has been installed")
+
+                # Execute pre start commands
+                shim_pre_start_commands = get_shim_pre_start_commands(runner_build)
+                run_pre_start_commands(client, shim_pre_start_commands)
+                logger.debug("The script for installing dstack has been executed")
+
+                # Run dstack-shim as a systemd service
+                run_shim_as_systemd_service(
+                    client=client,
+                    working_dir=DSTACK_WORKING_DIR,
+                    dev=settings.DSTACK_VERSION is None,
+                )
+
+                # Get host info
+                host_info = get_host_info(client, DSTACK_WORKING_DIR)
+                logger.debug("Received a host_info %s", host_info)
+
+        except ProvisioningError as e:
+            logger.exception("Failed remote instance deploy %s", e)
```

Comment:
I don't think it's a good idea to call logger.exception() on any error. For example, it would cause server exceptions being logged on dstack sky when dstack fails to connect to users' instances but users won't see any errors.

For expected errors such as when dstack fails to connect to instances, there should be no exceptions logged. Instead, the failed instance should have an associated error status / message, so that users can see them via CLI/UI.

Reply:
I deleted the logger.exception. There is currently no field to save the error message. I'll do it later.